python/lsst/daf/butler/core/databaseDict.py

```diff
@@ -0,0 +1,101 @@
+# This file is part of daf_butler.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from collections.abc import MutableMapping
+
+from .utils import doImport
+
+
+class DatabaseDict(MutableMapping):
+    """An abstract base class for dict-like objects backed by a database.
```

Comment:
"dict-like objects" implies something a bit simpler than what this generally is. It's not simple key/value pairs in a dict that are persisted to a database, it's more of a python model of a database table. I feel like we are close to reimplementing an astropy.table or afw table database serialization system.